ExampleMod/Content/Projectiles/ExampleWhipProjectile.cs

```diff
@@ -0,0 +1,168 @@
+			Projectile.height = 18;
+			Projectile.friendly = true;
+			Projectile.penetrate = -1;
+			Projectile.tileCollide = false;
+			Projectile.ownerHitCheck = true; // This prevents the projectile from hitting through solid tiles.
+			Projectile.extraUpdates = 1;
+			Projectile.usesLocalNPCImmunity = true;
+			Projectile.localNPCHitCooldown = -1;
+		}
+
+		private float Timer {
+			get => Projectile.ai[0];
+			set => Projectile.ai[0] = value;
+		}
+
+		private float ChargeTime {
+			get => Projectile.ai[1];
+			set => Projectile.ai[1] = value;
+		}
+
+		public override void AI() {
+			Player owner = Main.player[Projectile.owner];
+			Projectile.rotation = Projectile.velocity.ToRotation() + MathHelper.PiOver2; // Without PiOver2, the rotation would be off by 90 degrees counterclockwise.
+
+			Projectile.Center = Main.GetPlayerArmPosition(Projectile) + Projectile.velocity * Timer;
+			// Vanilla uses Vector2.Dot(Projectile.velocity, Vector2.UnitX) here. Dot Product returns the difference between two vectors.
+			// However, the use of UnitX basically turns it into a more complicated way of checking if the projectile's velocity is above or equal to zero on the X axis.
+			Projectile.spriteDirection = Projectile.velocity.X >= 0f ? 1 : -1;
+
+			// This statement handles charging. The maximum charge time is doubled to accommodate for the projectile's extra update. If you don't want this functionality, replace the statement with just Timer++;
+			if (owner.channel && ChargeTime < 120) {
+				ChargeTime++;
+			}
+			else {
+				Timer++;
+			}
+
+			Projectile.GetWhipSettings(Projectile, out float timeToFlyOut, out int _, out float _);
+
+			if (Timer >= timeToFlyOut || owner.itemAnimation <= 0) {
+				Projectile.Kill();
+				return;
+			}
+
+			owner.heldProj = Projectile.whoAmI;
+
+			// These two lines ensure that the timing of the owner's use animation is correct.
+			owner.itemAnimation = owner.itemAnimationMax - (int)(Timer / Projectile.MaxUpdates);
+			owner.itemTime = owner.itemAnimation;
+
+			if (Timer == timeToFlyOut / 2) {
+				// Plays a whipcrack sound at the tip of the whip.
+				List<Vector2> points = Projectile.WhipPointsForCollision;
+				Projectile.FillWhipControlPoints(Projectile, points);
+				SoundEngine.PlaySound(SoundID.Item153, points[points.Count - 1]);
+			}
+		}
+
+		public override void GetWhipSettings(Player player, ref float timeToFlyOut, ref int segments, ref float rangeMultiplier) {
+			segments = 20 + (int)ChargeTime / 12;
+			rangeMultiplier = 1 + (int)ChargeTime / 120f;
+		}
+
+		public override void OnHitNPC(NPC target, int damage, float knockback, bool crit) {
+			target.AddBuff(ModContent.BuffType<ExampleWhipDebuff>(), 240);
+			Main.player[Projectile.owner].MinionAttackTargetNPC = target.whoAmI;
+		}
+
+		// This method draws a line between all points of the whip, in case there's empty space between the sprites.
+		private void DrawLine(List<Vector2> list) {
+			Texture2D texture = TextureAssets.FishingLine.Value;
```

Comment:
Same thing as AI() -> Is this boilerplate code that every mod will have? Should this instead be a template method that mods call with a handful of parameters instead?

Reply:
In my opinion, it's best to have a normal 'ExampleX' that can be used as a template, and then 'ExampleAdavancedX' that reimplements some vanilla code instead of reusing it, if needed.